### ISSUE TITLE
how to slice a `Matrix`?

### ISSUE THREAD

**Issue by haydnv** [elftausend/custos-math#3]
Are there any ops provided for slicing or indexing? Or is this planned for the future?

For example:
```rust
let device = CPU::new();
let m = Matrix::<i32>::new(&device, (5, 8));
assert_eq!(m.slice(1..3, 2).dims(), (2, 1));
```

or maybe:
```rust
let slice = Matrix::from_slice_of(&m, 1..3, 2.into());
```

or even:
```rust
let coords = Matrix::slice_coords(1..3, 2.into);
let dest = Matrix::<i32>::new(2, 1);
device.assign_slice(&m, &coords, &mut dest);
```

**Comment by haydnv** [comment]
Thanks! I do need something that works on GPU devices (which IIUC `as_slice` does not). Let me know if I can help!

**Comment by elftausend** [comment]
You are right. `as_slice` only works with CPU `Matrix` (and if the `Matrix` was created with an OpenCL device on a physical device with unified memory). An invalid call to this function panics. In newer versions, this is prohibited during compile time.

However, slicing should also work on OpenCL to some extend as I found this: [SubBuffers].
`SubBuffers` should make slicing possible.
I searched a bit if there is an equivalent for CUDA but till now, I couldn't find one.

I am working on a WGPU backend and WGPU supports slicing as well: [BufferSlice]
I think, the inner workings / architecture of `BufferSlice` should also help implementing this in `custos` and `custos-math` 

[SubBuffers]: http://downloads.ti.com/mctools/esd/docs/opencl/memory/buffers.html#sub-buffers
[BufferSlice]: https://docs.rs/wgpu/latest/wgpu/struct.BufferSlice.html

**Comment by elftausend** [comment]
Yes, there will be a trait like `CopySlice` where a method will return something like `BufferSlice`. 

Nice. Thank you!
I will try this in a few days. (I am currently not on a CUDA capable machine.)

**Comment by haydnv** [comment]
Here's a PR for the changes in custos: https://github.com/elftausend/custos/pull/15

**Comment by haydnv** [comment]
Do you have any advice on how to test changes to custos with custos-math? I just get a bunch of compile errors when using the HEAD version of custos: https://gist.github.com/haydnv/f7b2d5847a8dbcc1cf5e68a3aee6a841

**Comment by elftausend** [comment]
Thank you for the PR.

I made a mistake where I merged an advanced branch into the main branch without doing the same with custos-math.
Currently, I am in the process of merging a newer branch into the main branch, which is compatible with the [custos-v6] branch of custos-math.

[custos-v6]: https://github.com/elftausend/custos-math/tree/custos-v6

**Comment by elftausend** [comment]
I merged the `custos-v6` branch with the main branch of custos-math.
PR: https://github.com/elftausend/custos-math/pull/4

**Comment by haydnv** [comment]
custos-math appears to be broken at HEAD: https://gist.github.com/haydnv/88fe281e14282db4794ba71dabcca9e9

**Comment by elftausend** [comment]
I added the option to disable the `cpu` feature and I did not fix the compilation errors of the tests if the `cpu` feature is disabled.
Now should it be possible to run the tests without any features, but it still recommended to add the `cpu` feature.
`cargo test --no-default-features --features cpu`

**Comment by haydnv** [comment]
Thanks! The tests for the `default` and `cpu` features now compile but none of the other feature configurations I tried work (`cuda`, `opencl`, etc)

**Comment by elftausend** [comment]
You are right. Forgot to automatically activate the `cpu` feature if only the `cuda` feature is used. Same problem as before. 
For `opencl`: I develop a lot on my machine with unified memory (laptop), and I overlooked trying to compile with manually disabling unified memory. (`CUSTOS_USE_UNIFIED=false cargo build --no-default-features --features=opencl`)

**Comment by haydnv** [comment]
Thanks! This is enough to un-block work on slicing. Just FYI `tests/test_rand.rs` still does not compile with `--features=cuda`).

**Comment by haydnv** [comment]
This should be fixed by #5

**Comment by haydnv** [comment]
Hi @elftausend, do you have an ETA for when these PRs will be merged? Is there anything else that needs to happen on my end?

**Comment by elftausend** [comment]
I will merge them in a few minutes.
Thank you very much for your work.
